### PR TITLE
introduce azure api bindings and ore azure subcommand

### PR DESCRIPTION
**Pull request by mischief**
for now, i've only added upload-blob until upstream azure-sdk-for-go is improved for our needs.

the azure subcommand only builds under go1.7 because of the azure api requirement of TLS renegotiations.

### CODE REVIEW

File: platform/api/azure/api.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	plog = capnslog.NewPackageLogger("github.com/coreos/mantle", "platform/api/azure")
+)
+
+type API struct {
+	client management.Client
+	opts   *Options
+}
+
+// New creates a new Azure client. If no publish settings file is provided or
+// can't be parsed, an anonymous client is created.
```
Comment (marineam) [inline platform/api/azure/api.go]: If the user expects to be logged in the failure should be fatal.

**Comment by marineam** [comment]
Other than the auth downgrade this looks good.

File: platform/api/azure/api.go (hunk 1 above)
Comment (mischief) [inline platform/api/azure/api.go]: unless i defer creation of the client until the api call, a user will always have to pass a publish settings file even if it isn't needed. it's not needed in the case of UploadBlob, which only uses storage account credentials.

Comment (mischief) [inline platform/api/azure/api.go]: is it more reasonable to always require both publishsettings file and storage credentials even when one set of credentials might not be used?

Comment (mischief) [inline platform/api/azure/api.go]: alternately, i could split the API into two parts.

Comment (crawford) [inline platform/api/azure/api.go]: The storage credentials can be derived from the publish settings.

Comment (marineam) [inline platform/api/azure/api.go]: Always using publish settings and deriving storage credentials from that sounds like a fairly user friendly scheme.

**Comment by mischief** [comment]
ptal

File: platform/api/azure/storage_mit.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	"github.com/Microsoft/azure-vhd-utils-for-go/upload"
+	"github.com/Microsoft/azure-vhd-utils-for-go/upload/metadata"
+	"github.com/Microsoft/azure-vhd-utils-for-go/vhdcore/common"
+	"github.com/Microsoft/azure-vhd-utils-for-go/vhdcore/diskstream"
+	"github.com/coreos/pkg/multierror"
+)
+
+const pageBlobPageSize int64 = 2 * 1024 * 1024
+
+// UploadBlob
```
Comment (crawford) [inline platform/api/azure/storage_mit.go]: Expand?

File: platform/api/azure/storage_mit.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+	return nil
+}
+
+// getBlobMetaData returns the custom metadata associated with a page blob which is set by createBlob method.
+// The parameter client is the Azure blob service client, parameter containerName is the name of an existing container
+// in which the page blob resides, parameter blobName is name for the page blob
+// This method attempt to fetch the metadata only if MD5Hash is not set for the page blob, this method panic if the
+// MD5Hash is already set or if the custom metadata is absent.
+//
+func getBlobMetaData(client storage.BlobStorageClient, containerName, blobName string) (*metadata.MetaData, error) {
```
Comment (crawford) [inline platform/api/azure/storage_mit.go]: getBlobMetadata

**Comment by crawford** [comment]
Why is that not vendored?

**Comment by crawford** [comment]
Looks reasonable.

**Comment by mischief** [comment]
@crawford because their version of "UploadBlob" is a gigantic function inside of package main: https://github.com/Microsoft/azure-vhd-utils/blob/master/vhdUploadCmdHandler.go

**Comment by crawford** [comment]
uhhhhhhhhhhh, okay. LGTM assuming the licensing is sorted out.

**Pull request merged**